### PR TITLE
react-table: port v7 types over from react-table repo

### PR DESCRIPTION
**Pull request by ggascoigne**
Since the v7 version of react-table has a completely incompatible
interface to v6, and is also no longer including types in its v7
package, this changeset ports over the v7 types and moves the v6 types
into a sub-folder.

The interface isn't ideal, it depends upon interface merging to be
useful.  If anyone can offer suggestions as to how to fix this without
falling into the recursive definition problems of generating types based
upon plugins which themselves change the types that they are being used
to define, please let me know.

more details can be found at https://gist.github.com/ggascoigne/646e14c9d54258e40588a13aabf0102d

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/tannerlinsley/react-table/blob/master/docs/api.md
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

regarding tslint:disable:no-unnecessary-generics
no-unnecessary-generics is disabled because many of these definitions are either used in a generic context or the signatures are required to match for declaration merging.

### CODE REVIEW

**Review by tannerlinsley — APPROVED** [review]

**Review by xPapla — APPROVED** [review]

**Review by kjaksik — APPROVED** [review]

**Comment by antonjoel82** [comment]
Hi, everyone:

First off, thank you all for your contributions on this! I'm new to Typescript, but hope to be able to contribute down the line.

Please let me know if this is not the place to discuss this so I can relocate it, but I'm currently getting errors when I attempt to utilize some of the hooks with `useTable`. I've looked over the test file based on the "kitchen sink" example, and don't see any major differences. Here's my initialization with the hooks: 
```JavaScript
const {
        getTableProps,
        getTableBodyProps,
        headerGroups,
        prepareRow,
        page,
        canPreviousPage,
        canNextPage,
        pageOptions,
        pageCount,
        gotoPage,
        nextPage,
        previousPage,
        setPageSize,
        state: { expanded, pageIndex, pageSize },
    } = useTable<Task>({ 
            columns,
            data: tasks,
            getSubRows: getSubTaskRows,
            initialState: { pageIndex: 2 }
        },         
        useExpanded,
        usePagination
    );
```

This results in the following Typescript errors:
- Property page, canPreviousPage, canNextPage, pageOptions, pageCount, gotoPage, nextPage, previousPage, setPageSize does not exist on type 'TableState<Task>'.ts(2339)
- Property 'expanded, pageIndex, pageSize' does not exist on type 'TableState<Task>'.ts(2339)

I've been able to temporarily work-around this issue by modifying the index.d.ts file and extending other interfaces. Is there something I should do differently to fix this in my own code? Thanks!

**Review by tannerlinsley — APPROVED** [review]

**Comment by rostero1** [comment]
Thanks for doing this.

I was trying to use these definitions with:  https://github.com/tannerlinsley/react-table/blob/master/examples/pagination-controlled/src/App.js, but ran into a few type errors around `initialState` and `manualPagination`.

**Comment by ggascoigne** [comment]
> Thanks for doing this.
> 
> I was trying to use these definitions with: https://github.com/tannerlinsley/react-table/blob/master/examples/pagination-controlled/src/App.js, but ran into a few type errors around `initialState` and `manualPagination`.

These types depend on declaration merging, see https://github.com/ggascoigne/react-table/blob/ts-kitchen-sink-conrolled/examples/kitchen-sink-controlled-ts/src/types/react-table-config.d.ts for an example.

**Comment by androsj** [comment]
@tannerlinsley is working so fast it's not giving this a break to merge xD

**Comment by ggascoigne** [comment]
Please don't merge this yet.  I'd really like the interface to stabilize for more than a day before calling this good.

File: types/react-table/index.d.ts
Comment (kizu) [inline types/react-table/index.d.ts]: This and other prop getters below now can be also just objects or arrays of objects instead of callbacks, and the callbacks can also return arrays of objects, see more there: https://github.com/tannerlinsley/react-table/issues/1749#issuecomment-564177091

File: types/react-table/index.d.ts
Comment (androsj) [inline types/react-table/index.d.ts]: To match type of selectedRowIds: `Set<string>`

```suggestion
    getRowId: (originalRow: D, relativeIndex: number) => string;
```

**Comment by androsj** [comment]
`autoResetSelectedRows` is missing from `UseRowSelectOptions`

File: types/react-table/index.d.ts
Comment (ggascoigne) [reply types/react-table/index.d.ts]: damnit, I thought I'd fixed this one :)

Thank you for taking the time to look at this and give your feedback.

File: types/react-table/index.d.ts
Comment (ggascoigne) [reply types/react-table/index.d.ts]: yeah, I just noticed this.

**Comment by Oleg-Ovcharenko** [comment]
In react-table we have setHiddenColumns  but not setColumnHidden

File: types/react-table/index.d.ts
Comment (tpict) [inline types/react-table/index.d.ts]: for prop getters like this, can we type the response as `object & { key: string }` in lieu of a more comprehensive interface?

Comment (kizu) [reply types/react-table/index.d.ts]: I think we could maybe declare separate types outside, one for each prop getter, like `TableHeaderProps`, `TableRowProps`, `TableCellProps` etc. and then use those in prop getters? Like

```ts
export interface TableCellProps {
  key?: React.Key;
}

// …

getCellProps: (props?: TableCellProps) => TableCellProps;
```

This would allow us to use declaration merging to extend those interfaces easily, adding the props that we would be using?

Comment (tpict) [reply types/react-table/index.d.ts]: Yeah absolutely. I just thought it was worth adding `key` at the very least since that needs to passed explicitly if you're enforcing `react/jsx-key` in your eslint config.

Comment (kizu) [reply types/react-table/index.d.ts]: Ah, yes, hmm, it would be nice to make the key mandatory in the output, but optional in the input!

File: types/react-table/index.d.ts
Comment (ggascoigne) [reply types/react-table/index.d.ts]: actions are typed now:

```tsx
export type ActionType = { type: string } & Record<string, any>;
```

File: types/react-table/index.d.ts
Comment (ggascoigne) [reply types/react-table/index.d.ts]: I refactored the types for all of the property getters, there might be a more compact way to have done it, but what's there is at least correct, I hope :)

Happy to take any suggestions about how to make them cleaner.

File: types/react-table/index.d.ts
Comment (ggascoigne) [reply types/react-table/index.d.ts]: done, separate types for each one.

File: types/react-table/index.d.ts
Comment (kizu) [reply types/react-table/index.d.ts]: Here is my PR for your PR :) https://github.com/ggascoigne/DefinitelyTyped/pull/1 — adds a generic that makes it much easier to create all the getters (and also fixes their signatures etc.)

**Comment by ggascoigne** [comment]
and by `intermediate`, I really meant `indeterminate` - bah.

File: types/react-table/index.d.ts
Comment (androsj) [inline types/react-table/index.d.ts]: This has the same default props as `TableToggleAllRowsSelectedProps `

File: types/react-table/index.d.ts
Comment (androsj) [inline types/react-table/index.d.ts]: These default props are identical to `TableToggleRowsSelectedProps` and `TableToggleAllRowsSelectedProps`.
How about making a common interface then? Something like `TableToggleCommonProps`

File: types/react-table/index.d.ts
Comment (androsj) [inline types/react-table/index.d.ts]: Should be optional`?` like the other autoResets ?

**Comment by andrewbranch** [comment]
It looks like this is still in flux and not ready to merge; is that right @ggascoigne?

**Comment by ggascoigne** [comment]
I can’t say that there will be no further changes before v7 is released, but the last few sets of changes have been around the plugin implementation interface rather than to the more public interfaces. I.e. the types for rc.12 would work pretty well with rc.15 unless you write your own plugins. 

So yes there might be changes, but hopefully they will be minor.

**Comment by andrewbranch** [comment]
You’ve clearly put a ton of work into this, had a lot of eyes on it, and already have several community approvals, so I’m happy to merge whenever you’re ready.

**Comment by JimmyMultani** [comment]
I would love to see this merged as is for now, with another PR created for when v7 is released. A lot of developers are stuck on beta.23 since it was the last version to actually have typescript defs in it.

**Comment by ggascoigne** [comment]
Ok. Turns out that there are a few more changes on the way. So we’d best hold off.

**Comment by ggascoigne** [comment]
@jimmymultani If you want to try these out early, I’d use patch-package to install them.

**Comment by ChrisB9** [comment]
Are you able to roughly estimate how long you need to hold it?

**Comment by ggascoigne** [comment]
@ChrisB9 Honestly I don't, I'm not really involved in the main lib, other than the odd small PR.  There are some more changes being worked on, I don't know how impactful they will be.

But I'm also not wedded to the idea that this PR doesn't get merged until v7 releases.  I think that it matches the current rc pretty well.

It seems that there is interest in merging it now and me opening a separate PR for any subsequent changes, I'm good with that.

@andrewbranch can you merge this when you get a chance.  Thank you.

**Pull request merged**
